bluepyentity/entity_definitions.py

```diff
@@ -0,0 +1,338 @@
+        if isinstance(item, list) and all(isinstance(i, accepted) for i in item):
+            return item
+
+        classlist = ", ".join(accepted)
+        raise TypeError(f"{classlist} or List[{classlist}] type expected")
+
+    @classmethod
+    def _custom_validator(cls, item):
+        return cls._as_list_of_accepted_types(item, cls.accepted)
+
+
+class ListOfStr(ListOfAccepted):
+    """Helper class to force strings to list of strings."""
+
+    accepted = {str}
+
+
+class DistributionConverter(ListOfAccepted):
+    """Convert Distribution entries to expected format"""
+
+    accepted = {str, dict}
+
+    @classmethod
+    def _custom_validator(cls, item):
+        items = cls._as_list_of_accepted_types(item, cls.accepted)
+
+        # item is already a dict or a str
+        def _distribution_dict(item):
+            return {"path": item} if isinstance(item, str) else item
+
+        return [_distribution_dict(i) for i in items]
+
+
+# Data Models
+
+
+class BaseModel(pydantic.BaseModel):
+    """Base model for the entities."""
+
+    type: str = None
+
+    class Config:
+
+        extra = pydantic.Extra.allow
+        arbitrary_types_allowed = True
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.type = self.type or self.__class__.__name__
+
+    @classmethod
+    def from_dict(cls, dict_):
+        """Create class from dict."""
+        # Could as well be from file
+        return cls.parse_obj(dict_)
+
+    def to_resource(self, forge):
+        """Convert the definition to a nexus-forge resource."""
+        data = self.get_formatted_definition(forge)
+
+        # These need to be handled separately after creating the resource
+        single_files = {"configuration", "template", "target"}
+        all_special_cases = single_files.union({"derivation", "image", "distribution"})
+        to_add = {k: data.pop(k, None) for k in all_special_cases}
+
+        # Resource needs to be a Dataset for it to have methods such as add_derivation.
+        resource = kgforge.core.Resource.from_json(data)
+        resource = kgforge.specializations.resources.Dataset.from_resource(forge, resource)
+
+        if to_add["derivation"] is not None:
+            derivation = _fetch(to_add["derivation"], forge)
+            resource.add_derivation(derivation)
+
+        if to_add["image"] is not None:
+            for item in to_add["image"]:
+                resource.add_files(item)
+
+        if to_add["distribution"] is not None:
+            for item in to_add["distribution"]:
+                resource.add_distribution(**item)
+
+        for file_key in single_files:
+            if to_add[file_key] is not None:
+                setattr(resource, file_key, forge.attach(path=to_add[file_key]))
+
+        return resource
+
+    def get_formatted_definition(self, forge):
+        """Return the formatted definition."""
+        res = {k: v for k, v in self.__dict__.items() if v is not None}
```

Comment:
Would it be possible to not use directly the internal dict but have a dedicated attribute for the resource-related attributes?

On a different note: you can use `vars(self)` instead of `self.__dict__`.

Reply:
Fetching a property doesn't imply it is readonly though. I am fine with both, tbh.
Sure, `dict()` is fine as well.